### ISSUE TITLE
Denoising Autoencoders target label

### ISSUE THREAD

**Issue by steveazzolin** [aapatel09/handson-unsupervised-learning#6]
In model 9-10 of chapter 8 I don't understand why the target label of the fitting phase is the noisy input. Doing this the autoencoders learns to reconstruct the noisy input, isn't it ?
Should instead be written in this way ?

```
history = model.fit(x=X_train_AE_noisy, y=X_train_AE,  #using the original input as target
                            epochs=10,
                            batch_size=32,
                            shuffle=True,
                            validation_data=(X_val_AE_noisy, X_val),
                            verbose=1)
```

Original code is:

```
history = model.fit(x=X_train_AE_noisy, y=X_train_AE_noisy,
                        epochs=num_epochs,
                        batch_size=batch_size,
                        shuffle=True,
                        validation_data=(X_train_AE, X_train_AE),
                        verbose=1)
```

I have also replaced 
`validation_data=(X_train_AE, X_train_AE)`
with
`validation_data=(X_val_AE_noisy, X_val)`
Where X_val is obtained by splitting the train set into 2 subsets. What do you think ?

Moreover why predictions are made on the noised test set ? I tried to predict the original test set.
Executing the code with these updates I got:

> Mean average precision over 10 runs:  0.6327761266275229
> Coefficient of variation over 10 runs:  0.018682160539079985
> [0.6243557484686052,
>  0.6346261618115446,
>  0.6111520781820431,
>  0.6294291611519081,
>  0.6387461341507094,
>  0.6258058349652793,
>  0.6550653293459515,
>  0.6473704778074738,
>  0.6260815948098096,
>  0.6351287455819044]

**Comment by aapatel09** [comment]
Hi, here I am simulating a case where we do not have the clean dataset at all. All we have to work with is the "noisy" data. The goal of the this autoencoder is to ignore the noise and instead learn to reconstruct just the core elements of X_train_AE. We don't want the autoencoder to learn to reconstruct perfectly X_train_AE_noisy at all. We want the autoencoder to learn to ignore the noise while it's learning. That's why we train on X_train_AE_noisy but we evaluate on X_train_AE. Please drop me a note if this is still unclear.

**Comment by steveazzolin** [comment]
Thank you for the reply,

> The goal of the this autoencoder is to ignore the noise and instead learn to reconstruct just the core elements of X_train_AE.

The concept is clear, but what is not clear to me is how we get this if the objective of the learning (the target of training) is the noisy input Itselfs. How do we "teach" to the Autoencoder what is the noise if It only sees noised data ?

**Comment by aapatel09** [comment]
Great question. By underfitting the autoencoder on the noisy data, the autoencoder learns the common "signal" but ignores the rare "noise". If the noise truly drowns out the signal in an overwhelming way, even underfitting the autoencoder won't cut it. But, in instances where the noise is present but not the majority of the data, underfitting works well in practice.

Does this underfitting clarify how using an autoencoder on a noisy dataset still achieves the end objective of learning the signal in the data but ignoring the noise?

**Comment by steveazzolin** [comment]
Yeah thank you! I have a more clear idea now.

Anyway, what about training the Autoencoder to reconstruct the original data (without noise) ? Could be an effective alternative ? (Just to complete my overview on this topic)
Obiously with the constraint that we have the un-noised data.

**Comment by aapatel09** [comment]
Yes, you are right. If you have clean data, training the autoencoder for
longer is better. You do not want to underfit the autoencoder, but you also
have to be careful about overfitting.

Please let me know if you have any questions.

Thanks!
Ankur
